packages/composites/proof-of-reserves/src/utils/addressValidator.ts

```diff
@@ -0,0 +1,209 @@
+import { utils } from 'ethers'
+import { Logger } from '@chainlink/ea-bootstrap'
+import type { AdapterResponse } from '@chainlink/types'
+import type { Validator } from '@chainlink/ea-bootstrap'
+
+type AddressObject = {
+  address: string
+  network?: string
+  chainId?: string
+}
+
+const indexerToNetwork: Record<string, string> = {
+  ada_balance: 'cardano',
+  eth_balance: 'ethereum',
+  bitcoin_json_rpc: 'bitcoin',
+  lotus: 'filecoin',
+}
+
+export const getValidAddresses = (
+  protocolOutput: AdapterResponse,
+  validator: Validator,
+): AdapterResponse => {
+  const validatedInput = { ...protocolOutput }
+  if (!parseBoolean(validator.validated.data.disableAddressValidation)) {
+    validatedInput.result = validateAddresses(
+      validator.validated.data.indexer,
+      validatedInput.result,
+    )
+  }
+  if (!parseBoolean(validator.validated.data.disableDuplicateAddressFiltering)) {
+    validatedInput.result = filterDuplicates(validatedInput.result)
+  }
+  validatedInput.data.result = validatedInput.result
+  return validatedInput
+}
+
+export const validateAddresses = (indexer: string, addresses: AddressObject[]): AddressObject[] => {
+  const validatedAddresses: AddressObject[] = []
+  for (const addressObj of addresses) {
+    const { address, network } = addressObj
+    let validatedAddress: string | undefined = undefined
+    const validationNetwork = network || indexerToNetwork[indexer]
+    switch (validationNetwork.toLowerCase()) {
+      case 'ethereum':
+        validatedAddress = getValidEvmAddress(address)
+        break
+      case 'bitcoin':
+        validatedAddress = getValidBtcAddress(address)
+        break
+      case 'dogecoin':
+        validatedAddress = getValidDogeAddress(address)
+        break
+      case 'cardano':
+        validatedAddress = getValidCardanoAddress(address)
+        break
+      case 'filecoin':
+        validatedAddress = getValidFilecoinAddress(address)
+        break
+      default:
+        Logger.debug(
+          `There is no address validation procedure defined for the "${network}" network.`,
+        )
+        validatedAddresses.push(addressObj)
+        break
+    }
+    if (validatedAddress) validatedAddresses.push({ ...addressObj, address: validatedAddress })
+  }
+  return validatedAddresses
+}
+
+/**
+ * Returns either a valid Ethereum-style address with a valid checksum
+ * or logs a warning and returns undefined
+ */
+const getValidEvmAddress = (address: string): string | undefined => {
+  try {
+    return utils.getAddress(address)
+  } catch (error) {
+    Logger.warn(
```

Comment:
We may want to consider two things:
- Turning this to log level error
- Erroring out completely

We could have the behavior connected to an environment variable.
It depends on how serious it is for an invalid address to be included. A warning log will be extremely easy for a node operator to miss.

Reply:
I asked Jonas about this and he preferred a log message.  If there is any debate, I would prefer to leave it as-is unless you think it is critical.

Reply:
To offer a different perspective, some NOPs have alerting even for warn logs afaik, so we do need to be careful that these are sporadic and if we add this it's for NOPs to be warned

Reply:
I would think that an invalid address being fed in is a fairly serious thing that should be immediately caught by nops.
I would advocate bumping this to an `error` level log.

@boxhock what do you think?
Though if we have enough nops that have warn log alerting it should be okay this way.